File: week06/5/prob1.cpp
// you can use includes, for example:
// #include <algorithm>
#include <iostream>
#include <vector>
#include <string>
using namespace std;

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

int solution(string &S) {
    // Implement your solution here
    int is_valid = 1;

    vector<char> stack;

    for (char c : S) {
        // open 이면 enqueue
        if (c=='(') {
            stack.push_back(')');
        } else if (c=='{') {
            stack.push_back('}');
        } else if (c=='[') {
            stack.push_back(']');
        // close 면 가능한 지 체크 후 dequeue
        } else {
            // empty 면 더 볼 것도 없이 invalid
            if (stack.empty()) {
                is_valid = 0;
                break;
            // empty 가 아니면 top 이 close 와 동류인지 확인하고
            // - 동류가 맞으면 dequeue
            } else if (stack.back()==c) {
                stack.pop_back();
            // - 동류가 아니면 invalid
            } else {
                is_valid = 0;
                break;
            }
        }
    }

    // for loop 를 끝까지 통과했다는 것은 잘못된 close 로 인한 invalid 는 없었다는 뜻.
    // 잘못된 open 즉, close 로 마무리되지 못한 open 도 체크해야 함.
    // 다시 말해, for loop 통과 후 stack 이 empty 되었는지 확인해야 함.
    if (!stack.empty()) {
        is_valid = 0;
    }
    return is_valid;
}

File: week06/5/prob1.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(S):
    # Implement your solution here

    invalid = 1         # 체크 하다가 이상함을 한 번이라도 발견하면 바로 out
    stack = []

    for c in S:
        # open 을 만나면 enqueue
        if c == '(':
            stack.append(')')
        elif c == '{':
            stack.append('}')
        elif c == '[':
            stack.append(']')

        # close 를 만나면 가능한지 체크 후 dequeue
        else:
            # stack 이 empty : 부적절한 close
            if len(stack) == 0:
                invalid = 0
                break
            # stack.top 이 동류가 아님 : 부적절한 close
            elif stack[-1] != c:
                invalid = 0
                break
            # 그 외 : 적절한 close : dequeue
            else:
                stack.pop()
    
    # for loop 은 close 가 적절했는지를 검사했음.
    # open 이 적절했는지 : stack 이 for loop 종료 후 empty 인지 확인해야함
    if invalid == 1 and len(stack) > 0:
        invalid = 0
    
    return invalid
